pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java

```diff
@@ -211,19 +211,12 @@ public FileVisitResult visitFile(final Path file,
 
     @Override
     public FileAnalysisListener startFileAnalysis(TextDocument file) {
-        String fileName = file.getPathId();
-        AnalysisResult analysisResult = updatedResultsCache.get(fileName);
-        if (analysisResult == null) {
-            analysisResult = new AnalysisResult(file.getCheckSum());
-        }
-        final AnalysisResult nonNullAnalysisResult = analysisResult;
+        final String fileName = file.getPathId();
 
         return new FileAnalysisListener() {
             @Override
             public void onRuleViolation(RuleViolation violation) {
-                synchronized (nonNullAnalysisResult) {
-                    nonNullAnalysisResult.addViolation(violation);
-                }
+                updatedResultsCache.get(fileName).addViolation(violation);
```

Comment:
`updatedResultsCache` is already a `ConcurrentHashMap`, and since there is a single listener per-file, being used by a single thread (by contract of `FileAnalysisListener`), the `addViolation` call is safe.